pkg/binlog/position.go

```diff
@@ -226,25 +226,43 @@ func (l Location) CloneWithFlavor(flavor string) Location {
 //   1 if point1 is bigger than point2
 //   0 if point1 is equal to point2
 //   -1 if point1 is less than point2
-func CompareLocation(location1, location2 Location) int {
-	if location1.GTIDSet != nil && len(location1.GTIDSet.String()) != 0 &&
-		location2.GTIDSet != nil && len(location2.GTIDSet.String()) != 0 {
-		contain1 := location1.GTIDSet.Contain(location2.GTIDSet)
-		contain2 := location2.GTIDSet.Contain(location1.GTIDSet)
+func CompareLocation(location1, location2 Location, cmpGTID bool) int {
+	if cmpGTID {
+		cmp, canCmp := CompareGTID(location1.GTIDSet, location2.GTIDSet)
+		if canCmp {
+			return cmp
+		}
+
+		// if can't compare by GTIDSet, then compare by position
+		log.L().Warn("gtidSet can't be compared, will compare by position", zap.Stringer("location1", location1), zap.Stringer("location2", location2))
+	}
+
+	return ComparePosition(location1.Position, location2.Position)
+}
+
+// CompareGTID returns:
+//   1, true if gSet1 is bigger than gSet2
+//   0, true if gSet1 is equal to gSet2
+//   -1, true if gSet1 is less than gSet2
+// but if can't compare gSet1 and gSet2, will returns 0, false
+func CompareGTID(gSet1, gSet2 gtid.Set) (int, bool) {
+	if gSet1 != nil && len(gSet1.String()) != 0 &&
+		gSet2 != nil && len(gSet2.String()) != 0 {
```

Comment:
Do we need to support compare a non-empty GTID Set with an empty GTID set? (may non-empty > empty? as we did in `gtid.Set.Contain`)

Reply:
I support it in f06dcd4